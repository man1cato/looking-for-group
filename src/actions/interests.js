import axios from 'axios';

const baseUrl = 'https://api.airtable.com/v0/appOY7Pr6zpzhQs6l';
const apiKey = 'keyzG8AODPdzdkhjG';


//GET_INTERESTS
export const getInterests = (interests) => ({
    type: 'GET_INTERESTS',
    interests
});

export const startGetInterests = () => {
    return async (dispatch) => {
        const response = await axios.get(`${baseUrl}/Interests?sort[0][field]=Interest&sort[0][direction]=asc&api_key=${apiKey}`);
        const interests = response.data.records.map((record) => ({
            id: record.id,
            name: record.fields.Interest
        }));
        dispatch(getInterests(interests));
    };
};

