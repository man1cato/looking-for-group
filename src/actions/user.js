import axios from 'axios';

const baseUrl = 'https://api.airtable.com/v0/appOY7Pr6zpzhQs6l';
const apiKey= 'keyzG8AODPdzdkhjG';


//SET_USER
export const setUser = (user = {}) => ({
    type: 'SET_USER',
    user
});

export const startSetUser = ({uid, email}) => {
    const filterFormula = `{Email}="${email}"`;
    return async (dispatch) => {
        try {
            //SEARCH AIRTABLE FOR USER'S EMAIL
            const response = await axios.get(`${baseUrl}/Users?filterByFormula=${filterFormula}&api_key=${apiKey}`);
            //IF EMAIL FOUND, GATHER PROFILE INFO
            if (response.data.records.length > 0) {
                const userRecord = response.data.records[0];
                const interest1 = userRecord.fields['#1 Interest'][0] ? userRecord.fields['#1 Interest'][0] : [];
                const interest2 = userRecord.fields['#2 Interest'][0] ? userRecord.fields['#2 Interest'][0] : [];
                const interest3 = userRecord.fields['#3 Interest'][0] ? userRecord.fields['#3 Interest'][0] : [];
                const additionalInterests = userRecord.fields["Add'l Interests"] ? userRecord.fields["Add'l Interests"] : [];
                const allInterests = [interest1, interest2, interest3, ...additionalInterests];
                const user = {
                    recordId: userRecord.id,
                    firstName: userRecord.fields['First Name'],
                    lastName: userRecord.fields['Last Name'],
                    email,
                    postalCode: userRecord.fields['Postal Code'],
                    birthYear: userRecord.fields['Birth Year'],
                    interest1,
                    interest2,
                    interest3,
                    additionalInterests,
                    allInterests,
                    availability: userRecord.fields.Availability,
                    area: userRecord.fields.Area[0],
                    groups: userRecord.fields.Groups
                };
                //IF FIREBASE ID NOT PRESENT IN RECORD, ADD TO RECORD
                if(!response.data.records[0].fields['Firebase ID']) {
                    axios.patch(`${baseUrl}/Users/${user.recordId}?api_key=${apiKey}`, {"fields": {"Firebase ID": uid} } );
                }
                dispatch(setUser(user));
                return user;
            } else {
                const postResponse = await axios.post(`${baseUrl}/Users?api_key=${apiKey}`, {"fields": {"Email": email, "Firebase ID": uid}});
                const user = {
                    recordId: postResponse.data.id,
                    email
                };
                dispatch(setUser(user));
                return user;
            }
        } catch (e) {
            throw new Error('Call to Airtable Users table failed');
        }
    };
};

//UPDATE_USER
export const updateUser = (user) => ({
    type: 'UPDATE_USER',
    user
});

export const startUpdateUser = (user) => {
    const fields = {
        "First Name": user.firstName,
        "Last Name": user.lastName,
        "Postal Code": user.postalCode,
        "Birth Year": Number(user.birthYear),
        "#1 Interest": [user.interest1],
        "#2 Interest": [user.interest2],
        "#3 Interest": [user.interest3],
        "Add'l Interests": user.additionalInterests,
        "Availability": user.availability
    };
    return (dispatch) => {
        if (user.recordId) {
            axios.patch(`${baseUrl}/Users/${user.recordId}?api_key=${apiKey}`, {"fields": fields});
        } else {
            axios.post(`${baseUrl}/Users?api_key=${apiKey}`, {"fields": {"Email": user.email, "Firebase ID": user.uid, ...fields}});
        }
        dispatch(updateUser(user));
    };
};


// //UPDATE_USER_AREA  
// export const updateUserArea = (area) => ({
//     type: 'UPDATE_USER_AREA',
//     area
// });

// export const startUpdateUserArea = (user, placeDetails) => {
//     // const userRecordId = user.recordId;
//     // const userInterests = user.allInterests;
//     return (dispatch) => {
//         updateAirtable(user, placeDetails);
//         dispatch(updateUser(user));
//     };
//     // return async (dispatch) => {
//     //     try {
//     //         const areaRecordId = await geolocateUser(userRecordId, placeDetails);
//     //         const usersGroups = await addUserToGroups(userRecordId, userInterests, areaRecordId);
//     //         // for (let groupRecordId of usersGroups) {                                        //3. UPDATE GROUP AVAILABILITY
//     //         //     const group = await axios.get(`${baseUrl}/Groups/${groupRecordId}?api_key=${apiKey}`);
//     //         //     await updateGroupAvailability(group, userAvailability);                                 
//     //         // }
//     //         dispatch(updateUser({...user, area: areaRecordId, groups: usersGroups}));
//     //     } catch (e) {
//     //         throw new Error('Error at addUserToGroups within startUpdateUserArea');
//     //     }
//     // };
// };
