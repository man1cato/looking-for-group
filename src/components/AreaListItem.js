import React from 'react';
import {Link} from 'react-router-dom';

const AreaListItem = ({id, name}) => (
    <Link className="list-item" to={`/areas/${id}`}>                             
        <div>
            <h3 className="list-item__title">{name}</h3>
        </div>
    </Link>
);


export default AreaListItem;