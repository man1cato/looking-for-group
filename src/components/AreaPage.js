import React from 'react';
import {connect} from 'react-redux';
import getAreas from '../selectors/areas';

export const AreaPage = (props) => (
    <div>
        <h1>Areas</h1>
        <div>
            Put areas here
        </div>
    </div>
);

// const mapStateToProps = (state) => {
//     const view = 'Alpha Test List';
//     return {
//         areas: getAreas(view)
//     };
// };

// export default connect(mapStateToProps)(AreaPage);
