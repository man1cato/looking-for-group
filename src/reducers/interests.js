//INTERESTS REDUCER
const interestsReducer = (state = [], action) => {
    switch (action.type) {
        case 'GET_INTERESTS':
            return action.interests;
        default:
            return state;
    }
};

export default interestsReducer;