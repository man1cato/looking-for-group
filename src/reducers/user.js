//USER REDUCER
const userReducer = (state = [], action) => {
    switch (action.type) {
        case 'SET_USER':
            return action.user;
        case 'UPDATE_USER':
            return action.user;
        case 'UPDATE_USER_AREA':
            return action.area;   
        case 'UPDATE_USERS_GROUPS':
            return action.groups;
        default:
            return state;
    }
};

export default userReducer;