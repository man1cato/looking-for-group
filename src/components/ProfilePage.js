import React from 'react';
import {connect} from 'react-redux';
import {startUpdateUser, startUpdateUserArea, startUpdateUsersGroups} from '../actions/user';
// import InterestSelector from './InterestSelector';
import filterAvailabilities from '../utils/filterAvailabilities';
import geolocateUser from '../utils/geolocateUser';
import addUserToGroups from '../utils/addUserToGroups';

export class ProfilePage extends React.Component {
    constructor(props) {
        super(props);
        this.state = {      //default state
            uid: props.user.uid,
            email: props.user.email,
            recordId: props.user.recordId,
            firstName: props.user.firstName ? props.user.firstName : '',
            lastName: props.user.lastName ? props.user.lastName : '',
            postalCode: props.user.postalCode ? props.user.postalCode : '',
            birthYear: props.user.birthYear ? props.user.birthYear : 1990,
            interest1: props.user.interest1 ? props.user.interest1 : '',
            interest2: props.user.interest2 ? props.user.interest2 : '',
            interest3: props.user.interest3 ? props.user.interest3 : '',
            additionalInterests: props.user.additionalInterests ? props.user.additionalInterests : [],
            availability: props.user.availability ? props.user.availability : '',
            area: props.user.area,
            groups: props.user.groups 
        };
    };
    onTextChange = (e) => {
        const name = e.target.name;
        const value = e.target.value;
        this.setState(() => ({ [name]: value }));
    };
    onBirthYearChange = (e) => {
        const birthYear = e.target.value;
        this.setState(() => ({ birthYear }));
    };
    onInterestChange = (e) => {
        const name = e.target.name;
        const value = e.target.value;
        this.setState(() => ({ [name]: value }));
    }
    onAvailabilityChange = (e) => {
        const checked = e.target.checked;
        const changedAvailability = e.target.name;
        if (checked) {
            this.setState(() => ({ availability: [changedAvailability, ...this.state.availability] }));
            console.log('checked after:',this.state.availability);
        } else {
            this.setState(() => ({ availability: this.state.availability.filter((availability) => availability !== changedAvailability) }));
            console.log('unchecked after:',this.state.availability);
        }
    }
    onSubmit = (e) => {
        e.preventDefault();
        if (this.state.postalCode !== this.props.user.postalCode) {
            
            const scriptBlock = document.createElement('div');          //CREATE <DIV> TO HOLD TEMPORARY DATA
            scriptBlock.id = 'scriptBlock';
            document.body.appendChild(scriptBlock);
            
            const scriptNode = document.createElement('script');
            scriptNode.innerHTML = `getPlaceDetails('${this.state.recordId}','${this.state.postalCode}')`;  //MAKE CALL TO GOOGLE PLACES THEN CREATE <DIV> WITH RESPONSE INSIDE
            document.getElementById('scriptBlock').appendChild(scriptNode);
            
            setTimeout(() => {
                const placeDetails = JSON.parse(document.getElementById('placeDetails').textContent);
                this.props.startUpdateUserArea(this.state, placeDetails);                      //UPDATE USER'S AREA
                document.body.removeChild(scriptBlock);                         //DELETE <DIV> BLOCK TO CLEAR DATA
            }, 2000);
        } 
        
        setTimeout(() => {this.props.startUpdateUser(this.state)}, 6000);                                 //UPDATE USER'S PROFILE
        
        const snackbar = document.getElementById("snackbar");
        snackbar.className = "snackbar--show";
        setTimeout(() => { snackbar.className = snackbar.className.replace("snackbar--show", ""); }, 2000);
        
        setTimeout(() => {this.props.history.push('/')}, 8000);
        
    };
    render() {
        return (
            <div className="content-container">
            {console.log('Render ProfilePage:', this.state)}
                <div>
                    <h2>User Profile</h2>
                </div> 
                <div>
                    <form className="form" onSubmit={this.onSubmit}>
                        <div>
                            Name: <input
                                className="text-input"
                                type="text"
                                name="firstName"
                                value={this.state.firstName}
                                onChange={this.onTextChange}
                            />
                            <input 
                                className="text-input"
                                type="text"
                                name="lastName"
                                value={this.state.lastName}
                                onChange={this.onTextChange}
                            />
                        </div>
                        <div>
                            Birth Year: <input 
                                className="text-input"
                                type="number" 
                                name="birthYear" 
                                min="1900" 
                                max="2020"
                                value={this.state.birthYear}
                                onChange={this.onBirthYearChange}
                            />
                        </div>
                        <div>
                            Postal Code: <input 
                                className="text-input"
                                type="text" 
                                name="postalCode"
                                value={this.state.postalCode}
                                onChange={this.onTextChange}
                            />
                        </div>
                        <div>
                            #1 Interest: <select 
                                className="select" 
                                name="interest1" 
                                defaultValue={this.state.interest1}
                                onChange={this.onInterestChange}
                            >
                                <option key="0" value=""></option>          //blank option
                                {this.props.interests.map((interest) => (<option key={interest.id} value={interest.id}>{interest.name}</option>) )}
                            </select>
                        </div>
                        <div>
                            #2 Interest: <select 
                                className="select" 
                                name="interest2" 
                                defaultValue={this.state.interest2}
                                onChange={this.onInterestChange}
                            >
                                <option key="0" value=""></option>          //blank option
                                {this.props.interests.map((interest) => (<option key={interest.id} value={interest.id}>{interest.name}</option>) )}
                            </select>
                        </div>
                        <div>
                            #3 Interest: <select 
                                className="select" 
                                name="interest3" 
                                defaultValue={this.state.interest3}
                                onChange={this.onInterestChange}
                            >
                                <option key="0" value=""></option>          //blank option
                                {this.props.interests.map((interest) => (<option key={interest.id} value={interest.id}>{interest.name}</option>) )}
                            </select>
                        </div>
                        <div>
                            <table className="table">
                                <thead>
                                    <tr>
                                        <th>Availability:</th>
                                        <td>Mon</td>
                                        <td>Tue</td>
                                        <td>Wed</td>
                                        <td>Thu</td>
                                        <td>Fri</td>
                                        <td>Sat</td>
                                        <td>Sun</td>
                                    </tr>
                                </thead>
                                <tbody>
                                    <tr>
                                        <td className="table__first-column">Morning (8am-12pm)</td>
                                        {filterAvailabilities(this.props.availabilities, 'Morning').map((availability) => (
                                            <td key={availability.recordId}>
                                                <input 
                                                    type="checkbox" 
                                                    key={availability.recordId} 
                                                    name={availability.recordId} 
                                                    defaultChecked={this.state.availability.indexOf(availability.recordId) > -1 && true}
                                                    onChange={this.onAvailabilityChange}
                                                />
                                            </td> 
                                        ))}
                                    </tr>
                                    <tr>
                                        <td className="table__first-column">Afternoon (12pm-5pm)</td>
                                        {filterAvailabilities(this.props.availabilities, 'Afternoon').map((availability) => (
                                            <td key={availability.recordId}>
                                                <input 
                                                    type="checkbox" 
                                                    key={availability.recordId}
                                                    name={availability.recordId} 
                                                    defaultChecked={this.state.availability.indexOf(availability.recordId) > -1 && true}
                                                    onChange={this.onAvailabilityChange}
                                                />
                                            </td> 
                                        ))}
                                    </tr>
                                    <tr>
                                        <td className="table__first-column">Evening (5pm-10pm)</td>
                                        {filterAvailabilities(this.props.availabilities, 'Evening').map((availability) => (
                                            <td key={availability.recordId}>
                                                <input 
                                                    type="checkbox" 
                                                    key={availability.recordId} 
                                                    name={availability.recordId} 
                                                    defaultChecked={this.state.availability.indexOf(availability.recordId) > -1 && true}
                                                    onChange={this.onAvailabilityChange}
                                                />
                                            </td> 
                                        ))}
                                    </tr>
                                    <tr>
                                        <td className="table__first-column">Late Night (10pm-2am)</td>
                                        {filterAvailabilities(this.props.availabilities, 'Late Night').map((availability) => (
                                            <td key={availability.recordId}>
                                                <input 
                                                    type="checkbox" 
                                                    key={availability.recordId} 
                                                    name={availability.recordId} 
                                                    defaultChecked={this.state.availability.indexOf(availability.recordId) > -1 && true}
                                                    onChange={this.onAvailabilityChange}
                                                />
                                            </td> 
                                        ))}
                                    </tr>
                                </tbody>
                            </table>
                        </div>
                        <div>
                            <button id="profile-button" className="button">Save Profile</button>
                        </div>
                        <div id="map"></div>
                    </form>
                </div>
                <div className="snackbar" id="snackbar">Profile has been updated! Redirecting to dashboard...</div>
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    user: state.user,
    interests: state.interests,
    availabilities: state.availabilities
});

const mapDispatchToProps = (dispatch) => ({
    startUpdateUser: (user) => dispatch(startUpdateUser(user)),
    startUpdateUserArea: (recordId, placeDetails) => dispatch(startUpdateUserArea(recordId, placeDetails))
});

export default connect(mapStateToProps, mapDispatchToProps)(ProfilePage);


// e.preventDefault();
//         if (this.state.postalCode !== this.props.user.postalCode) {
//             const script = document.createElement('script');
//             script.innerHTML = `getPlaceDetails('${this.state.recordId}','${this.state.postalCode}')`;
//             document.body.appendChild(script);
//             setTimeout(() => {
//                 const placeDetails = JSON.parse(document.getElementById('placeDetails').textContent);
//                 this.props.startUpdateUserArea(this.state, placeDetails);                      //UPDATE USER'S AREA
//             },2000);
            
//             setTimeout(() => console.log('State after startUpdateUserArea:', this.state), 5000);
//         }
//         const snackbar = document.getElementById("snackbar");
//         snackbar.className = "snackbar--show";
//         setTimeout(() => { snackbar.className = snackbar.className.replace("snackbar--show", "") }, 2000);
        
//         setTimeout(() => this.props.startUpdateUser(this.state), 5000);                                 //UPDATE USER'S PROFILE
//         // setTimeout(() => {this.props.history.push('/')}, 5000);
//         setTimeout(() => console.log('Dispatched state:', this.state), 8000); 