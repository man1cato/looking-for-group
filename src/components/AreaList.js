import React from 'react';
import {connect} from 'react-redux';
import AreaListItem from './AreaListItem';


export const AreaList = (props) => (
    <div>
        <div className="page-header">
            <div className="content-container">
                <h2 className="page-header__title">Areas</h2>
                <span>Select the area in which you live.</span>
            </div>
        </div>
        <div className="content-container">
            <div className="list-body">
                {
                    props.areas.length === 0 ? (
                        <div className="list-item list-item--message">No areas</div>
                    ) : (
                        props.areas.map((area) => <AreaListItem key={area.id} {...area} /> )    
                    )
                }
            </div>
        </div>
    </div>
);


// const mapStateToProps = (state) =>  state.areas;

const mapStateToProps = (state) => ({ areas: state.areas });

export default connect(mapStateToProps)(AreaList);
