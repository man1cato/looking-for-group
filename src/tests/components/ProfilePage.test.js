import React from 'react';
import {shallow} from 'enzyme';
import {ProfilePage} from '../../components/ProfilePage';
import user from '../fixtures/user';
import interests from '../fixtures/interests';



test('should render ProfilePage correctly', () => {
    const startUpdateUser = jest.fn();
    const wrapper = shallow(
        <ProfilePage
            startUpdateUser={startUpdateUser}
            user={user}
            interests={interests}
        />);
    expect(wrapper).toMatchSnapshot();
});

test('should call all utils for valid form submission', () => {
    const getPlaceDetails = jest.fn();
    const geolocateUser = jest.fn();
    const wrapper = shallow(<ProfilePage user={user} onSubmit={onSubmitSpy} />);
    wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
    });
    expect(getPlaceDetails).toHaveBeenCalledWith(user.recordId, user.postalCode);
    expect(geolocateUser).toHaveBeenCalledWith();
});



// test('should handle startUpdateUser', () => {
//     wrapper.find('ProfilePage').prop('onSubmit')(user);
//     expect(history.push).toHaveBeenLastCalledWith('/');
//     expect(startUpdateUser).toHaveBeenLastCalledWith(user);
// });

