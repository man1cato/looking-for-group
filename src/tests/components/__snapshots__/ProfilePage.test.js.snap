// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ProfilePage correctly 1`] = `undefined`;
